File: RotateKtimes.java
public class RotateKtimes {
    //TC=O(n) SC=O(1)
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        if(nums==null||nums.length==0)return;
        if(k>=n)k=k%n;
        reverse(nums,0,n-1);
        reverse(nums,0,k-1);
        reverse(nums,k,n-1);

    }

    private void reverse(int[] nums,int l,int r){
        while(l<r){
            swap(nums,l,r);
            l++;
            r--;
        }
    }

    private void swap(int[] nums,int i,int j){
        if(i!=j){
            nums[i] = nums[i] + nums[j];
            nums[j] = nums[i] - nums[j];
            nums[i] = nums[i] - nums[j];
        }
    }
}

File: Rotate_k_times.py
"""
189. Rotate Array
Time Complexity - O(n)
Space Complexity - O(1)
So we reversed first len(nums)-k elements then reversed last k elements and revered the entire list to get output"""
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """ 
        k = k % len(nums)
        self.reverse(nums,0,len(nums)-k)
        self.reverse(nums,len(nums)-k,len(nums))
        self.reverse(nums,0,len(nums))
        return nums
            
    def reverse(self,array,start,end):
        end -= 1
        while(start < end):
            temp = array[start]
            array[start] = array[end]
            array[end] = temp
            start += 1
            end -= 1
            
            
        